### PR TITLE
[data]: Rename types file from .d.ts to .ts

### PR DESCRIPTION
**Pull request by dmsnell**
## Description

While working on #35869 I discovered that the module where we keep some
top-level type definitions is named with the `.d.ts` suffix as if it
were a manually-maintained types definition file.

It's capable of being a normal TypeScript module though, and my IDE was
having a bit of trouble with the `.d.ts` suffix - not because that was
wrong, but I think because it was different than what the IDE expected.

In this patch all we're doing is renaming the file in order to smooth
this over and clean up later type-related patches that will occur to
the data package. Updating `tsconfig.json` was necessary becuase of the
rename so that TypeScript will appropriately find and use these types.

## How has this been tested?
There should be no change in the bundled JavaScript code. There should be
no functional or visual changes.

❓ Could this impact any third-party projects dependent on `@wordpress/data`?
We need to verify that this generates the same type definitions for the package
as it previously did.

## Types of changes
Renaming a module holding type definitions.

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. 
- [x] My code follows the accessibility standards. 
- [x] I've tested my changes with keyboard and screen readers. 
- [x] My code has proper inline documentation. 
- [x] I've included developer documentation if appropriate. 
- [x] I've updated all React Native files affected by any refactorings/renamings in this PR (please manually search all `*.native.js` files for terms that need renaming or removal).

### CODE REVIEW

**Review by sirreal — APPROVED** [review]
This makes sense, thanks 👍

File: packages/data/tsconfig.json [hunk 1]
```diff
@@ -17,5 +17,6 @@
 		"src/redux-store/metadata/**/*",
 		"src/promise-middleware.js",
 		"src/utils",
+	  	"src/*.ts"
```
Comment (sirreal) [inline packages/data/tsconfig.json:20]: This could probably be `src/**/*.ts`, I expect TypeScript files anywhere are expected to be included 🙂

**Review by sarayourfriend — APPROVED** [review]
LGTM

File: packages/data/tsconfig.json (hunk 1 above)
Comment (dmsnell) [reply packages/data/tsconfig.json:20]: I'll update that when I actually start using `.ts` files in there. How does that sound?

**Comment by dmsnell** [comment]
I've rebuilt the project and it appears like the generated `types.d.ts` file is identical

**Pull request merged**